src/Data/V/Linear/Internal/V.hs

```diff
@@ -61,115 +62,74 @@ structure.
-elim xs f =
-  case caseNat @n of
-    Left Refl -> consumeV xs f
-    Right Refl -> elimS (split xs) f
+    next'# :: 1 <= n => V n a -> (# a, V (n - 1) a #)
+    next'# (V xs) = (# Vector.head xs, V (Vector.tail xs) #)
+
+next :: 1 <= n => V n a %1 -> (a, V (n - 1) a)
+next = Unsafe.toLinear next'
   where
-    elimS :: 1 <= n => (# a, V (n - 1) a #) %1 -> FunN n a b %1 -> b
-    elimS (# x, xs' #) g = case predNat @n of
-      Dict -> elim xs' (expandFunN @n @a @b g x)
-
--- XXX: This can probably be improved a lot.
-make :: forall n a. KnownNat n => FunN n a (V n a)
-make = case caseNat @n of
-  Left Refl -> V Vector.empty
-  Right Refl -> contractFunN @n @a @(V n a) prepend
-    where
-      prepend :: a %1 -> FunN (n - 1) a (V n a)
-      prepend t = case predNat @n of
-        Dict -> continue @(n - 1) @a @(V (n - 1) a) (cons t) (make @(n - 1) @a)
+    next' :: 1 <= n => V n a -> (a, V (n - 1) a)
+    next' (V xs) = (Vector.head xs, V (Vector.tail xs))
+
+-- Backward compatility
+-- split :: 1 <= n => V n a %1 -> (# a, V (n - 1) a #)
+-- split = next#
+
+type Elim :: Nat -> Type -> Type -> Type -> Constraint
+class (n ~ Arity b f) => Elim n a b f | n a b -> f, f b -> n where
+  elim :: f %1 -> V n a %1 -> b
+
+instance Elim 0 a b b where
+  elim b v =
+    consume v & \case
+      () -> b
+  {-# INLINE elim #-}
+
+instance (1 <= n, n ~ Arity b (a %1 -> f), Elim (n - 1) a b f) => Elim n a b (a %1 -> f) where
+  elim g v =
+    next v & \case
+      (a, v') -> elim (g a) v'
+  {-# INLINE elim #-}
 
 cons :: forall n a. a %1 -> V (n - 1) a %1 -> V n a
 cons = Unsafe.toLinear2 $ \x (V v) -> V (Vector.cons x v)
 
-continue :: forall n a b c. KnownNat n => (b %1 -> c) %1 -> FunN n a b %1 -> FunN n a c
-continue = case caseNat @n of
-  Left Refl -> id
-  Right Refl -> \f t -> contractFunN @n @a @c (continueS f (expandFunN @n @a @b t))
-    where
-      continueS :: (KnownNat n, 1 <= n) => (b %1 -> c) %1 -> (a %1 -> FunN (n - 1) a b) %1 -> (a %1 -> FunN (n - 1) a c)
-      continueS f' x a = case predNat @n of Dict -> continue @(n - 1) @a @b f' (x a)
-
-iterate :: forall n a. (KnownNat n, 1 <= n) => (a %1 -> (a, a)) -> a %1 -> V n a
-iterate dup init =
-  go @n init
-  where
-    go :: forall m. (KnownNat m, 1 <= m) => a %1 -> V m a
-    go a =
-      case predNat @m of
-        Dict -> case caseNat @(m - 1) of
-          Prelude.Left Refl ->
-            case pr1 @m Refl of
-              Refl ->
-                (make @m @a :: a %1 -> V m a) a
-          Prelude.Right Refl ->
-            dup a & \(a', a'') ->
-              a' `cons` go @(m - 1) a''
-
-    -- An unsafe cast to prove the simple equality.
-    pr1 :: forall k. 0 :~: (k - 1) -> k :~: 1
-    pr1 Refl = Unsafe.coerce Refl
+-- TODO: fails because Elim and make are not dual with the current implementation
+-- fromReplicator :: forall n a. KnownNat n => Replicator a %1 -> V n a
+-- fromReplicator = Replicator.elim (make @n)
```

Comment:
I see your problem, this should be solvable though. You just need to ask GHC to verify an additional fact (which happens to always be true).

```suggestion
fromReplicator :: forall n a. (KnownNat n, Replicator.Elim n a (V n a) (FunN n a (V n a))  => Replicator a %1 -> V n a
fromReplicator = Replicator.elim (make @n)
```